packages/contracts/contracts/B.Protocol/BAMM.sol

```diff
@@ -0,0 +1,284 @@
+        collateralValue = SP.getDepositorCollateralGain(address(this));
+        if (address(collateralERC20) == address(0)) {
+            collateralValue += address(this).balance;
+        } else {
+            collateralValue += collateralERC20.balanceOf(address(this));
+        }
+    }
+
+    function deposit(uint256 thusdAmount) external {        
+        // update share
+        uint256 thusdValue = SP.getCompoundedTHUSDDeposit(address(this));
+        uint256 collateralValue = getCollateralBalance();
+
+        uint256 price = fetchPrice();
+        require(collateralValue == 0 || price > 0, "deposit: chainlink is down");
+
+        uint256 totalValue = thusdValue + collateralValue * price / PRECISION;
+
+        // this is in theory not reachable. if it is, better halt deposits
+        // the condition is equivalent to: (totalValue = 0) ==> (total = 0)
+        require(totalValue > 0 || total == 0, "deposit: system is rekt");
+
+        uint256 newShare = PRECISION;
+        if(total > 0) newShare = total * thusdAmount / totalValue;
+
+        // deposit
+        require(thusdToken.transferFrom(msg.sender, address(this), thusdAmount), "deposit: transferFrom failed");
+        SP.provideToSP(thusdAmount);
+
+        // update LP token
+        mint(msg.sender, newShare);
+
+        emit UserDeposit(msg.sender, thusdAmount, newShare);        
+    }
+
+    function withdraw(uint256 numShares) external {
+        uint256 thusdValue = SP.getCompoundedTHUSDDeposit(address(this));
+        uint256 collateralValue = getCollateralBalance();
+
+        uint256 thusdAmount = thusdValue * numShares / total;
+        uint256 collateralAmount = collateralValue * numShares / total;
+
+        // this withdraws thusdn and eth
+        SP.withdrawFromSP(thusdAmount);
+
+        // update LP token
+        burn(msg.sender, numShares);
+
+        // send thusd and eth
+        if(thusdAmount > 0) thusdToken.transfer(msg.sender, thusdAmount);
+        emit UserWithdraw(msg.sender, thusdAmount, collateralAmount, numShares);        
+        if(collateralAmount == 0) {
+            return;
+        }
+
+        if (address(collateralERC20) == address(0)) {
+            // ETH
+            (bool success, ) = msg.sender.call{ value: collateralAmount }(""); // re-entry is fine here
+            require(success, "withdraw: sending ETH failed");
+        } else {
+            // ERC20
+            bool success = collateralERC20.transfer(msg.sender, collateralAmount);
+            require(success, "withdraw: sending collateral failed");
+        }
+    }
+
+    function addBps(uint256 n, int bps) internal pure returns(uint) {
+        require(bps <= 10000, "reduceBps: bps exceeds max");
+        require(bps >= -10000, "reduceBps: bps exceeds min");
+
+        return n * uint256(10000 + bps) / 10000;
+    }
+
+    function compensateForTHusdDeviation(uint256 ethAmount) public view returns(uint256 newEthAmount) {
```

Comment:
Not sure about this one, but these differences on how thUSD is written were intentional, or are a Typo?, e.g in 'THusd'?

Reply:
I renamed with preserving case how it was in initial BAMM contract, so how it would look better? `compensateForthUSDDeviation`?